sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/FieldBuilder.java

```diff
@@ -0,0 +1,259 @@
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(GeoPoint.class, DataType.EDM_GEOGRAPHY_POINT);
+    }
+
+    private static final List<Class<?>> UNSUPPORTED_TYPES = Arrays.asList(Byte.class,
+        Character.class,
+        char.class,
+        Float.class,
+        float.class,
+        Short.class,
+        short.class);
+
+    /**
+     * Creates a collection of {@link Field} objects corresponding to the properties of the type supplied.
+     *
+     * @param modelClass The class for which fields will be created, based on its properties.
+     * @param <T> The generic type of the model class.
+     * @return A collection of fields.
+     */
+    public static <T> List<Field> build(Class<T> modelClass) {
+        ClientLogger logger = new ClientLogger(FieldBuilder.class);
+        return build(modelClass, new Stack<>(), logger);
+    }
+
+    /**
+     * Recursive class to build complex data type.
+     *
+     * @param curClass Current class to be built.
+     * @param classChain A class chain from {@code modelClass} to prior of {@code curClass}.
+     * @param logger {@link ClientLogger}.
+     * @return A list of {@link Field} that curClass is built to.
+     */
+    private static List<Field> build(Class<?> curClass, Stack<Class<?>> classChain, ClientLogger logger) {
+        if (classChain.contains(curClass)) {
+            logger.warning(String.format("There is circular dependencies %s, %s", classChain, curClass));
+            return null;
+        }
+        if (classChain.size() > MAX_DEPTH) {
+            throw logger.logExceptionAsError(new RuntimeException(
+                "The dependency graph is too deep. Please review your schema."));
+        }
+        classChain.push(curClass);
+        List<java.lang.reflect.Field> classFields = Arrays.asList(curClass.getDeclaredFields());
+        List<Field> searchFields = classFields.stream()
+            .filter(classField -> !classField.isAnnotationPresent(FieldIgnore.class))
+            .map(classField -> buildField(classField, classChain, logger))
+            .collect(Collectors.toList());
+        classChain.pop();
+        return searchFields;
+    }
+
+    private static Field buildField(java.lang.reflect.Field classField, Stack<Class<?>> classChain,
+        ClientLogger logger) {
+        Type type = classField.getGenericType();
+
+        if (isSupportedNoneParameterizedType(type)) {
+            return buildNoneParameterizedType(classField, logger);
+        }
+        if (isArrayOrList(type)) {
+            return buildCollectionField(classField, classChain, logger);
+        }
+        List<Field> childFields = build((Class<?>) type, classChain, logger);
+        Field searchField = convertToBasicSearchField(classField, logger);
+        searchField.setFields(childFields);
+        return searchField;
+    }
+
+    private static boolean isSupportedNoneParameterizedType(Type type) {
```

Comment:
Is this helper method needed?

Reply:
It is used for distinguish whether it is Supported type or complex one. It is necessary to have in top level checking.

Reply:
The reason I think it isn't needed as it is checking something we can do without calling another method in a single line, just replacing all instances of this being called with:

```java
SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(type);
```

Reply:
Does not need to be removed now.

Reply:
Make sense. Will simplify